Improve setup-env script to allow selection machine and distro

kekiefer opened the pull request:
This change lays the groundwork for the support of multiple distros
that will eventually be contained within this repository.

It also makes the script a bit more user friendly and aids in the
discovery of different machine types and distro names.

kekiefer committed:
Improve setup-env script to allow selection machine and distro

This change lays the groundwork for the support of multiple distros
that will eventually be contained within this repository.

It also makes the script a bit more user friendly and aids in the
discovery of different machine types and distro names.

madisongh commented on setup-env:
```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+
+usage()
+{
+    echo -e "
```
Maybe use `cat <<EOF` here instead of echo with a multiline string?

madisongh commented on setup-env:
```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+done
+
+if [ "$(whoami)" = "root" ]; then
+    echo "ERROR: do not use the BSP as root. Exiting..."
```
Not actually exiting here.  Also, I've run into issues in the past with systems that don't have `whoami`.  Maybe just use `$USER` here?  I believe that's a POSIX shell standard variable.  I've also used `$(id -u) -eq 0`.

madisongh commented on setup-env:
```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
-    [ -n "$BUILDDIR" ] && cd $BUILDDIR
-TEMPLATECONF=`cat $BUILDDIR/conf/templateconf.cfg`
+
+export TEMPLATECONF=${TEMPLATECONF:-$OEROOT/meta-testdistro/conf}
```
There is a `.templateconf` file that can be sourced for this setting.

madisongh commented on setup-env:
```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+. $OEROOT/openembedded-core/oe-init-build-env $CWD/$1 > /dev/null
+
+# if conf/local.conf not generated, no need to go further
+if [ ! -e conf/local.conf ]; then
```
Shouldn't this be checking for `-n "$BUILDDIR" -a -e "$BUILDDIR/conf/local.conf"`?

Oh, I see you've changed to invoking the full `oe-init-build-env`, so I guess the current working directory will have been changed to $BUILDDIR if it's successful.  But if there's an error, and you happen to have a `conf/local.conf` under your original working directory... ?

madisongh commented on setup-env:
```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+        -e "s,SDKMACHINE ?=.*,SDKMACHINE ?= '$SDKMACHINE',g" \
+        -e "s,DISTRO ?=.*,DISTRO ?= '$DISTRO',g" \
+        -e "s,PACKAGE_CLASSES ?=.*,PACKAGE_CLASSES ?= '$PACKAGE_CLASSES',g" \
+        -i conf/local.conf
```
What if those variables *aren't* set in the environment? I'm not sure this is a good idea, at least not without some kind of message to indicate that the config is being changed (and even then I'm not necessarily convinced).

I do see a message below with some of the above settings mentioned, but not all.

kekiefer replied:
Note that this is only ever called when creating a new conf, this will never update an existing conf.

In reviewing this, I guess `PACKAGE_CLASSES` shouldn't be in there, since we're not providing a default or forcing the user to set it, so I'll take that out. It's not much use in the tegra setup anyhow. All the other variables would either set by the user or provided defaults in this script.

madisongh commented on setup-env:
```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+            echo "Linking $s/site.conf to conf/site.conf"
+            ln -s $s/site.conf conf
+        fi
+    done
```
This too looks strange to me. I'm OK with adding something like this (assuming I'm understanding what it's for), but having it be per-distro would be nice.

kekiefer replied:
I've seen this done in other distros and I like the pattern. Usually these kinds of settings are applicable across multiple builds -- to use a common download directory or point a few different builds at the same pr server. You'll see later on in the script that if it's not picked up globally, it will get pulled out of templateconf.

madisongh commented on setup-env:
```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+fi
+
+TDROOT=`readlink -f "$CWD/$BMETA"`
+export TDROOT
```
What's this for?

kekiefer replied:
This was exported in the original version of the your `setup-env` script, so I assumed you were using it. If not, I'll take it out.

kekiefer replied:
Although, now that I'm making some other changes here, I'm realizing that OEROOT gets unset by the `oe-init-build-env`, but I need this path later on. So even if it doesn't need to exported, I'm going move this up so I can use it.

madisongh commented on setup-env:
```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+    clean_up
+    return 1
+fi
+
```
Suggested addition here:
```
(cd $CWD; git submodule sync > /dev/null 2>&1)
(cd $CWD; git submodule status | grep '^-' | awk '{print $2}' | xargs git submodule update --init --)
```
One common problem I've run into with my submodule-based distros is people forgetting to initialize the submodules on first clone, or forgetting to update the submodules after a `git pull`.   I never got round to adding that here, though, because I never forget. :)

madisongh commented on setup-env:
```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+
+if [ "$build_dir_setup_enabled" = "true" ] && [ -z "$DISTRO" ]; then
+    usage
+    echo -e "ERROR: You must set DISTRO when creating a new build directory."
```
Why is DISTRO required? I'd rather have a sane, common default, if we can.

madisongh commented on setup-env:
```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
 fi
+
+if [ -z "$SDKMACHINE" ]; then
+    SDKMACHINE='x86_64'
```
I'd rather just set the default in our `local.conf.template` to `x86_64`.  I'm not sure I'd ever want to support `i686` for SDKMACHINE.

madisongh reviewed (commented):
Generally OK with the intent here, but have some comments in-line.

This will take some getting used to for me, and I'll have to update my automated builders to deal with the changes before this goes in.

kekiefer committed:
setup-env: use cat <<EOF for help

Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446653215

kekiefer committed:
setup-env: fix protection against usage as root

Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446654949

kekiefer committed:
setup-env: let oe-init-build-env use the .templateconf

Also don't export TDROOT, but keep it around since OEROOT gets unset

Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446655343
Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446656857

kekiefer committed:
setup-env: be more precise about location of local.conf for checks

Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446655878

kekiefer committed:
setup-env: automatically update submodules when run

Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446658378

kekiefer committed:
setup-env: don't set PACKAGE_CLASSES or SDKMACHINE

Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446658671
Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446656265

kekiefer committed:
setup-env: use testdistro as default if DISTRO not set

Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446658497

kekiefer commented:
I think that the site.conf coming from the template in meta-tegra as a fallback may have been broken in the first version. Let me know if this works better for you.

Otherwise, all changes made. Let me know and I'll squash it.

kekiefer committed:
local.conf.sample: update SDKMACHINE to be x86_64 by default

madisongh commented on setup-env:
```diff
@@ -4,53 +4,199 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+ARGS=$(getopt --options $SHORTOPTS  \
+  --longoptions $LONGOPTS --name $PROGNAME -- "$@" )
+# Print the usage menu if invalid options are specified
+if [ $? != 0 -o $# -lt 1 ]; then
```
Can we have <build-dir> default to `build`, as it does for poky & OE-Core, instead of requiring an argument?

madisongh commented on setup-env:
```diff
@@ -4,53 +4,199 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+Usage: MACHINE=<machine> DISTRO=<distro> source $PROGNAME <build-dir>
+Usage:                                   source $PROGNAME <build-dir>
+    <machine>    machine name
+    <distro>     distro name
```
Maybe mention that DISTRO= is optional now.

madisongh commented on setup-env:
```diff
@@ -4,53 +4,199 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+LONGOPTS="help"
+
+ARGS=$(getopt --options $SHORTOPTS  \
+  --longoptions $LONGOPTS --name $PROGNAME -- "$@" )
```
What do you think about adding options for specifying the machine name and distro, instead of requiring them to be specified as environment variables?

kekiefer replied:
Yeah, I can do that with all these options. I was thinking about adding this later anyhow to support setting local.conf options for things like the bootloader type, so these basic features might as well be in there too, especially if they're mostly going to have default options.

So right now, the only required option is going to be the machine name.

What do you think about using the machine name in the default build directory, like using `build-${MACHINE}` as the default?

madisongh replied:
>  What do you think about using the machine name in the default build directory, like using build-${MACHINE} as the default?

I'd rather stick with just `build`.  We're supposed to be able to support building for different machines in the same workspace, and what I've seen with defaulting to separate workspaces per machine is that errors creep in that break that support.

kekiefer committed:
setup-env: move machine and distro to getopt and builddir default

Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446906446
Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446906848
Resolves https://github.com/madisongh/test-distro/pull/7#discussion_r446907321

kekiefer commented:
Ok, everything is moved to getopt now.

madisongh commented:
Thanks.  Give me a couple of days to work through the changes I need to make and test them.

madisongh commented:
I managed to do some testing in the background while watching some ELC sessions today, and it looks OK.  With the modifications you made, turns out it was just a minor tweak to the autobuilder config to make it work.  Thanks.
